libtransact/src/scheduler/serial/core.rs

```diff
@@ -156,40 +175,147 @@ impl SchedulerCore {
             return Ok(());
         }
 
-        if self.current_batch.is_some() {
+        if self.current_txn.is_some() {
             return Ok(());
         }
 
-        let mut shared = self
-            .shared_lock
-            .lock()
-            .expect("scheduler shared lock is poisoned");
+        if self.current_batch.is_none() {
+            let mut shared = self
+                .shared_lock
+                .lock()
+                .expect("scheduler shared lock is poisoned");
+
+            if let Some(unscheduled_batch) = shared.pop_unscheduled_batch() {
+                self.txn_queue = unscheduled_batch.batch().transactions().to_vec();
+                self.current_batch = Some(unscheduled_batch);
+            } else {
+                return Ok(());
+            }
+        }
 
-        if let Some(pair) = shared.pop_unscheduled_batch() {
-            if pair.batch().transactions().len() != 1 {
-                unimplemented!("can't handle more than one transaction per batch");
+        let transaction = self.txn_queue.pop().ok_or_else(|| {
+            CoreError::InternalError(
+                "There are no transactions left in the queue for the current batch".into(),
```

Comment:
This error could include the batch id, and could be shorter. "No transactions left for the current batch: {}"

Reply:
Done